context/src/junit/parser.rs

```diff
@@ -0,0 +1,588 @@
+    Reader,
+};
+use thiserror::Error;
+
+use super::date_parser::JunitDateParser;
+
+const TAG_REPORT: &[u8] = b"testsuites";
+const TAG_TEST_SUITE: &[u8] = b"testsuite";
+const TAG_TEST_CASE: &[u8] = b"testcase";
+const TAG_TEST_CASE_STATUS_FAILURE: &[u8] = b"failure";
+const TAG_TEST_CASE_STATUS_ERROR: &[u8] = b"error";
+const TAG_TEST_CASE_STATUS_SKIPPED: &[u8] = b"skipped";
+const TAG_TEST_RERUN_FAILURE: &[u8] = b"rerunFailure";
+const TAG_TEST_RERUN_ERROR: &[u8] = b"rerunError";
+const TAG_TEST_RERUN_FLAKY_FAILURE: &[u8] = b"flakyFailure";
+const TAG_TEST_RERUN_FLAKY_ERROR: &[u8] = b"flakyError";
+const TAG_TEST_RERUN_STACK_TRACE: &[u8] = b"stackTrace";
+const TAG_SYSTEM_OUT: &[u8] = b"system-out";
+const TAG_SYSTEM_ERR: &[u8] = b"system-err";
+
+pub mod extra_attrs {
+    pub const FILE: &str = "file";
+    pub const FILEPATH: &str = "filepath";
+    pub const LINE: &str = "line";
+}
+
+#[derive(Error, Debug, Copy, Clone, PartialEq, Eq)]
+pub enum JunitParseError {
+    #[error("could not parse report name")]
+    ReportName,
+    #[error("no reports found")]
+    ReportNotFound,
+    #[error("multiple reports found")]
+    ReportMultipleFound,
+    #[error("report end tag found without start tag")]
+    ReportStartTagNotFound,
+    #[error("could not parse test suite name")]
+    TestSuiteName,
+    #[error("test suite found without a report found")]
+    TestSuiteReportNotFound,
+    #[error("test suite end tag found without start tag")]
+    TestSuiteStartTagNotFound,
+    #[error("could not parse test case name")]
+    TestCaseName,
+    #[error("test case found without a test suite found")]
+    TestCaseTestSuiteNotFound,
+    #[error("test case end tag found without start tag")]
+    TestCaseStartTagNotFound,
+    #[error("test case status found without a test case found")]
+    TestCaseStatusTestCaseNotFound,
+    #[error("test rerun found without a test case found")]
+    TestRerunStartTagNotFound,
+    #[error("test rerun end tag found without start tag")]
+    TestRerunTestCaseNotFound,
+    #[error("system out is empty")]
+    SystemOutEmpty,
+    #[error("system err is empty")]
+    SystemErrEmpty,
+    #[error("stack trace is empty")]
+    StackTraceEmpty,
+}
+
+#[derive(Debug, Clone)]
+enum Text {
+    SystemOut(Option<String>),
+    SystemErr(Option<String>),
+    StackTrace(Option<String>),
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct JunitParser {
+    date_parser: JunitDateParser,
+    errors: Vec<JunitParseError>,
+    reports: Vec<Report>,
+    current_report: Option<Report>,
+    current_test_suite: Option<TestSuite>,
+    current_test_case: Option<TestCase>,
+    current_test_rerun: Option<TestRerun>,
+    current_text: Option<Text>,
+}
+
+impl JunitParser {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn errors(&self) -> &Vec<JunitParseError> {
+        &self.errors
+    }
+
+    pub fn reports(&self) -> &Vec<Report> {
+        &self.reports
+    }
```

Comment:
I'm using [`quick-junit`](https://docs.rs/quick-junit/latest/quick_junit/) since it is what [`cargo-nextest`](https://nexte.st/) uses and is full-featured including test reruns. Most JUnit parsers do not include all the data that `quick-junit` does making it ideal for our use-cases. Unfortunately, there's not a parser / deserializer built into `quick-junit` (although we could contribute this back), so I've created this one here.